File: OpenBurnBarCore/Sources/OpenBurnBarKernel/SharedModels/MobilePulseWindowPolicy.swift
import Foundation

/// Pulse timeline scopes used by the mobile hero. Source oracle:
/// `PulseWindowMetricBuilder` in `OpenBurnBarMobile/Views/Pulse/PulseWindowMetrics.swift`.
public enum MobilePulseTimelineScope: String, Sendable, CaseIterable {
    case minute
    case hour
    case day
    case week
    case month
}

public struct MobilePulseUsageEvent: Sendable, Equatable {
    public let startMs: Int64
    public let endMs: Int64
    public let tokens: Int
    public let costUsd: Double

    public init(startMs: Int64, endMs: Int64, tokens: Int, costUsd: Double) {
        self.startMs = startMs
        self.endMs = endMs
        self.tokens = tokens
        self.costUsd = costUsd
    }

    /// iOS oracle: `max(startTime, endTime)`. Sync `updatedAt` is not an event time.
    public var eventDateMs: Int64 { max(startMs, endMs) }
}

public struct MobilePulseRollupTotals: Sendable, Equatable {
    public let requests: Int
    public let tokens: Int
    public let costUsd: Double

    public static let zero = MobilePulseRollupTotals(requests: 0, tokens: 0, costUsd: 0)

    public init(requests: Int, tokens: Int, costUsd: Double) {
        self.requests = requests
        self.tokens = tokens
        self.costUsd = costUsd
    }
}

public struct MobilePulseWindowResult: Sendable, Equatable {
    public let total: MobilePulseRollupTotals
    public let trailing: MobilePulseRollupTotals?

    public init(total: MobilePulseRollupTotals, trailing: MobilePulseRollupTotals?) {
        self.total = total
        self.trailing = trailing
    }
}

public enum MobilePulseLoadPresentation: String, Sendable, Equatable {
    case loading
    case failed
    case empty
    case live
    case staleRefreshFailed = "stale-refresh-failed"

    /// Failed / empty / stale-refresh must never render as a live $0 hero.
    public var looksLikeLiveZero: Bool { self == .live }
}

/// Shared Pulse/Burn window math. Authority: iOS `PulseWindowMetricBuilder`.
public enum MobilePulseWindowPolicy {
    public static let dayWindowMs: Int64 = 24 * 60 * 60 * 1_000

    public static func metrics(
        scope: MobilePulseTimelineScope,
        rollups: [String: MobilePulseRollupTotals],
        usages: [MobilePulseUsageEvent],
        nowMs: Int64
    ) -> MobilePulseWindowResult {
        switch scope {
        case .minute:
            return liveMetrics(usages: usages, startMs: nowMs - 60_000, endMs: nowMs, trailing: rollups["7d"])
        case .hour:
            return liveMetrics(usages: usages, startMs: nowMs - 3_600_000, endMs: nowMs, trailing: rollups["7d"])
        case .day:
            return liveMetrics(usages: usages, startMs: nowMs - dayWindowMs, endMs: nowMs, trailing: rollups["7d"])
        case .week:
            return MobilePulseWindowResult(total: rollups["7d"] ?? .zero, trailing: rollups["30d"])
        case .month:
            return MobilePulseWindowResult(total: rollups["30d"] ?? .zero, trailing: rollups["90d"])
        }
    }

    /// Firestore query start: floor `now - 24h` to the hour in `timeZoneIdentifier`.
    public static func liveQueryStartMs(nowMs: Int64, timeZoneIdentifier: String) -> Int64 {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: timeZoneIdentifier)
            ?? TimeZone(secondsFromGMT: 0)!
        let now = Date(timeIntervalSince1970: Double(nowMs) / 1_000)
        let rollingStart = now.addingTimeInterval(-Double(dayWindowMs) / 1_000)
        let components = calendar.dateComponents([.year, .month, .day, .hour], from: rollingStart)
        let floored = calendar.date(from: components) ?? rollingStart
        return Int64((floored.timeIntervalSince1970 * 1_000).rounded())
    }

    public static func loadPresentation(
        isLoading: Bool,
        failed: Bool,
        hasCachedData: Bool
    ) -> MobilePulseLoadPresentation {
        if isLoading && !hasCachedData { return .loading }
        if failed { return hasCachedData ? .staleRefreshFailed : .failed }
        if !hasCachedData { return .empty }
        return .live
    }

    public static func currencyHero(costUsd: Double) -> String {
        max(0, costUsd).formatAsCost()
    }

    public static func tokensHero(tokens: Int) -> String {
        max(0, tokens).formatAsTokenVolume()
    }

    public static func quotaDedupKey(provider: String, accountId: String?, accountLabel: String?) -> String {
        let providerKey = provider.lowercased()
        let accountKey = accountId?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
            ?? accountLabel?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
            ?? "provider-level"
        return "\(providerKey)::\(accountKey.lowercased())"
    }

    public static func sortQuotaKeys(_ keys: [String]) -> [String] {
        keys.sorted { lhs, rhs in
            lhs.localizedCaseInsensitiveCompare(rhs) == .orderedAscending
        }
    }

    public static func pulseCost(costUsd: Double, costUSD: Double, cost: Double) -> Double {
        let raw = [costUsd, costUSD, cost].first { $0 != 0 } ?? 0
        return max(0, raw)
    }

    public static func pulseTokens(
        totalTokens: Int,
        inputTokens: Int,
        outputTokens: Int,
        cacheCreationTokens: Int,
        cacheReadTokens: Int,
        reasoningTokens: Int
    ) -> Int {
        let billed = max(0, inputTokens) + max(0, outputTokens) + max(0, cacheCreationTokens)
            + max(0, cacheReadTokens) + max(0, reasoningTokens)
        let raw = totalTokens != 0 ? totalTokens : billed
        return max(0, raw)
    }

    private static func liveMetrics(
        usages: [MobilePulseUsageEvent],
        startMs: Int64,
        endMs: Int64,
        trailing: MobilePulseRollupTotals?
    ) -> MobilePulseWindowResult {
        let rows = usages.filter { event in
            let at = event.eventDateMs
            return at >= startMs && at <= endMs
        }
        return MobilePulseWindowResult(
            total: MobilePulseRollupTotals(
                requests: rows.count,
                tokens: rows.reduce(0) { $0 + max(0, $1.tokens) },
                costUsd: rows.reduce(0) { $0 + max(0, $1.costUsd) }
            ),
            trailing: trailing
        )
    }
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}

File: android/app/src/main/java/com/openburnbar/data/policy/MobilePulseWindowPolicy.kt
package com.openburnbar.data.policy

import java.util.Calendar
import java.util.Locale
import java.util.TimeZone
import kotlin.math.max

/**
 * Pulse timeline scopes. Source oracle:
 * `PulseWindowMetricBuilder` in
 * `OpenBurnBarMobile/Views/Pulse/PulseWindowMetrics.swift`.
 */
enum class MobilePulseTimelineScope {
    MINUTE,
    HOUR,
    DAY,
    WEEK,
    MONTH,
}

data class MobilePulseUsageEvent(
    val startMs: Long,
    val endMs: Long,
    val tokens: Long,
    val costUsd: Double,
) {
    /** iOS oracle: `max(startTime, endTime)`. Sync `updatedAt` is not an event time. */
    val eventDateMs: Long get() = max(startMs, endMs)
}

data class MobilePulseRollupTotals(
    val requests: Int,
    val tokens: Long,
    val costUsd: Double,
) {
    companion object {
        val ZERO = MobilePulseRollupTotals(0, 0, 0.0)
    }
}

data class MobilePulseWindowResult(
    val total: MobilePulseRollupTotals,
    val trailing: MobilePulseRollupTotals?,
)

enum class MobilePulseLoadPresentation(val wire: String) {
    LOADING("loading"),
    FAILED("failed"),
    EMPTY("empty"),
    LIVE("live"),
    STALE_REFRESH_FAILED("stale-refresh-failed"),
    ;

    val looksLikeLiveZero: Boolean get() = this == LIVE
}

/** Shared Pulse/Burn window math. Authority: iOS `PulseWindowMetricBuilder`. */
object MobilePulseWindowPolicy {
    const val DAY_WINDOW_MS: Long = 24L * 60L * 60L * 1_000L
    const val HOUR_WINDOW_MS: Long = 60L * 60L * 1_000L
    const val MINUTE_WINDOW_MS: Long = 60L * 1_000L
    private const val COST_ZERO_EPSILON = 1e-9
    private const val COST_CENT_THRESHOLD = 0.01
    private const val TOKEN_BILLION = 1_000_000_000
    private const val TOKEN_MILLION = 1_000_000
    private const val TOKEN_THOUSAND = 1_000

    fun metrics(
        scope: MobilePulseTimelineScope,
        rollups: Map<String, MobilePulseRollupTotals>,
        usages: List<MobilePulseUsageEvent>,
        nowMs: Long,
    ): MobilePulseWindowResult = when (scope) {
        MobilePulseTimelineScope.MINUTE ->
            liveMetrics(usages, nowMs - MINUTE_WINDOW_MS, nowMs, rollups["7d"])
        MobilePulseTimelineScope.HOUR ->
            liveMetrics(usages, nowMs - HOUR_WINDOW_MS, nowMs, rollups["7d"])
        MobilePulseTimelineScope.DAY ->
            liveMetrics(usages, nowMs - DAY_WINDOW_MS, nowMs, rollups["7d"])
        MobilePulseTimelineScope.WEEK ->
            MobilePulseWindowResult(rollups["7d"] ?: MobilePulseRollupTotals.ZERO, rollups["30d"])
        MobilePulseTimelineScope.MONTH ->
            MobilePulseWindowResult(rollups["30d"] ?: MobilePulseRollupTotals.ZERO, rollups["90d"])
    }

    fun liveQueryStartMs(nowMs: Long, timeZoneIdentifier: String): Long {
        val tz =
            if (timeZoneIdentifier.isBlank()) {
                TimeZone.getTimeZone("GMT")
            } else {
                TimeZone.getTimeZone(timeZoneIdentifier)
            }
        val calendar = Calendar.getInstance(tz)
        calendar.timeInMillis = nowMs - DAY_WINDOW_MS
        calendar.set(Calendar.MINUTE, 0)
        calendar.set(Calendar.SECOND, 0)
        calendar.set(Calendar.MILLISECOND, 0)
        return calendar.timeInMillis
    }

    fun loadPresentation(isLoading: Boolean, failed: Boolean, hasCachedData: Boolean): MobilePulseLoadPresentation = when {
        isLoading && !hasCachedData -> MobilePulseLoadPresentation.LOADING
        failed && hasCachedData -> MobilePulseLoadPresentation.STALE_REFRESH_FAILED
        failed -> MobilePulseLoadPresentation.FAILED
        !hasCachedData -> MobilePulseLoadPresentation.EMPTY
        else -> MobilePulseLoadPresentation.LIVE
    }

    fun currencyHero(costUsd: Double): String = formatAsCost(max(0.0, costUsd))

    fun tokensHero(tokens: Long): String = formatAsTokenVolume(max(0L, tokens))

    fun quotaDedupKey(provider: String, accountId: String?, accountLabel: String?): String {
        val providerKey = provider.lowercase()
        val accountKey =
            accountId?.trim()?.takeIf { it.isNotEmpty() }
                ?: accountLabel?.trim()?.takeIf { it.isNotEmpty() }
                ?: "provider-level"
        return "$providerKey::${accountKey.lowercase()}"
    }

    fun sortQuotaKeys(keys: List<String>): List<String> = keys.sortedWith(String.CASE_INSENSITIVE_ORDER)

    fun pulseCost(costUsd: Double, costUSD: Double, cost: Double): Double {
        val raw = sequenceOf(costUsd, costUSD, cost).firstOrNull { it != 0.0 } ?: 0.0
        return max(0.0, raw)
    }

    fun pulseTokens(totalTokens: Int, inputTokens: Int, outputTokens: Int, cacheCreationTokens: Int, cacheReadTokens: Int, reasoningTokens: Int): Long {
        val billed =
            max(0, inputTokens).toLong() + max(0, outputTokens) + max(0, cacheCreationTokens) +
                max(0, cacheReadTokens) + max(0, reasoningTokens)
        val raw = if (totalTokens != 0) totalTokens.toLong() else billed
        return max(0L, raw)
    }

    private fun liveMetrics(usages: List<MobilePulseUsageEvent>, startMs: Long, endMs: Long, trailing: MobilePulseRollupTotals?): MobilePulseWindowResult {
        val rows = usages.filter { it.eventDateMs in startMs..endMs }
        return MobilePulseWindowResult(
            total = MobilePulseRollupTotals(
                requests = rows.size,
                tokens = rows.sumOf { max(0L, it.tokens) },
                costUsd = rows.sumOf { max(0.0, it.costUsd) },
            ),
            trailing = trailing,
        )
    }

    private fun formatAsCost(value: Double): String {
        if (value < COST_ZERO_EPSILON) return "$0.00"
        return if (value < COST_CENT_THRESHOLD) {
            "$" + "%.4f".format(Locale.US, value)
        } else {
            "$" + "%.2f".format(Locale.US, value)
        }
    }

    private fun formatAsTokenVolume(tokens: Long): String {
        val magnitude = tokens.toDouble()
        return when {
            tokens >= TOKEN_BILLION -> "%.2fB".format(Locale.US, magnitude / TOKEN_BILLION)
            tokens >= TOKEN_MILLION -> "%.2fM".format(Locale.US, magnitude / TOKEN_MILLION)
            tokens >= TOKEN_THOUSAND -> "%.1fK".format(Locale.US, magnitude / TOKEN_THOUSAND)
            else -> tokens.toString()
        }
    }
}
